File: notes/07_fp/15_env_model/shapes.ts
type Shape = {
    area: () => number;
    perimeter: () => number;
};

function createCircle(r: number): Shape {
    return {
        radius: r,
        area: () => Math.PI * r * r,
        perimeter: () => 2 * Math.PI * r
    } as Shape;
}

function createSquare(a: number): Shape {
    return {
        side: a, 
        area: () => a * a,
        perimeter: () => 4 * a
    } as Shape;
}

const circle : Shape = createCircle(1);
const square : Shape = createSquare(1);

console.log(circle.area()); 
console.log(square.area()); 

File: notes/07_fp/15_env_model/shapes.cpp
#include <iostream>
#include <functional>
#include <numbers>

struct Shape
{
    std::function<double()> area;
    std::function<double()> perimeter;
};

Shape createSquare(double a)
{
    return {[a]()->double{return a*a;},
            [a]()->double{return 4*a;}};
}

Shape createCircle(double r)
{
    return {[r]()->double{return std::numbers::pi*r*r;},
            [r]()->double{return 2*std::numbers::pi*r;}};
}

int main()
{
    Shape circle = createCircle(1),
          square = createSquare(1);
    std::cout << circle.area() << std::endl;
    std::cout << square.area() << std::endl;
}

